pypeit/coadd3d.py

```diff
@@ -0,0 +1,1772 @@
+    @classmethod
+    def get_instance(cls, spec2dfiles, opts, spectrograph=None, par=None, det=1, overwrite=False,
+                     show=False, debug=False):
+        """
+        Instantiate the subclass appropriate for the provided spectrograph.
+
+        The class to instantiate must match the ``pypeline``
+        attribute of the provided ``spectrograph``, and must be a
+        subclass of :class:`CoAdd3D`; see the parent class
+        instantiation for parameter descriptions.
+
+        Returns:
+            :class:`CoAdd3D`: One of the subclasses with
+            :class:`CoAdd3D` as its base.
+        """
+
+        return next(c for c in cls.__subclasses__()
+                    if c.__name__ == (spectrograph.pypeline + 'CoAdd3D'))(
+                        spec2dfiles, opts, spectrograph=spectrograph, par=par, det=det, overwrite=overwrite,
+                        show=show, debug=debug)
+
+    def __init__(self, files, opts, spectrograph=None, par=None, det=None, overwrite=False,
+                 show=False, debug=False):
+        """
+
+        Args:
+            files (:obj:`list`):
+                List of all spec2D files
+            opts (:obj:`dict`):
+                Options associated with each spec2d file
+            spectrograph (:obj:`str`, :class:`~pypeit.spectrographs.spectrograph.Spectrograph`, optional):
+                The name or instance of the spectrograph used to obtain the data.
+                If None, this is pulled from the file header.
+            par (:class:`~pypeit.par.pypeitpar.PypeItPar`, optional):
+                An instance of the parameter set.  If None, assumes that detector 1
+                is the one reduced and uses the default reduction parameters for the
+                spectrograph (see
+                :func:`~pypeit.spectrographs.spectrograph.Spectrograph.default_pypeit_par`
+                for the relevant spectrograph class).
+            det (:obj:`int`_, optional):
+                Detector index
+            overwrite (:obj:`bool`, optional):
+                Overwrite the output file, if it exists?
+            show (:obj:`bool`, optional):
+                Show results in ginga
+            debug (:obj:`bool`, optional):
+                Show QA for debugging.
+
+        """
+        self.spec2d = files
+        self.numfiles = len(files)
+        self.opts = opts
+        self.overwrite = overwrite
+
+        # Check on Spectrograph input
+        if spectrograph is None:
+            with fits.open(files[0]) as hdu:
+                spectrograph = hdu[0].header['PYP_SPEC']
+
+        if isinstance(spectrograph, str):
+            self.spec = load_spectrograph(spectrograph)
+            self.specname = spectrograph
+        else:
+            # Assume it's a Spectrograph instance
+            self.spec = spectrograph
+            self.specname = spectrograph.name
+
+        # Grab the parset, if not provided
+        if par is None:
```

Comment:
self.par = self.spec.default_pypeit_par() if par is None else par

Reply:
Done